### PR TITLE
Added functionality to support responses to "dud" and "sim" subjects

### PR DESCRIPTION
**Pull request by hughdickinson**
…specified using metadata in the manifest. Also provided functionality to validate the spatial click coordinates of a "sim" classification based on X, Y and tolerance values, also provided as metadata in a manifest.
# Review Checklist
- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] Did you deploy a staging branch?
## Optional
- [ ] If it's a new component, is it in ES6? Is it clear of warnings from ESLint?
- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?

### CODE REVIEW

**Comment by camallen** [comment]
staged at https://hughdickinson-metadata-based-feedback.pfe-preview.zooniverse.org/

**Comment by camallen** [comment]
@hughdickinson what's your testing project url?

**Comment by hughdickinson** [comment]
A project that demonstrates the new functionality can be found [here](https://hughdickinson-metadata-based-feedback.pfe-preview.zooniverse.org/projects/hughdickinson/testing-metadata-based-feedback).

**Comment by srallen** [comment]
Looks like maybe a bad rebase happened. Some grid tool functionality looks like it was removed when it shouldn't have been. L20 and L409-413 need to be added back in.

Are we ok with using `#` as a part of the metadata header to id the messages? We're telling project builders in the UI of the subject uploader that any header using `#` stays hidden from volunteers.

Also, I think this may a good opportunity to rework the summary part of the classifier and this makes sense to me as being in its own component rather than adding more to the classifier index.cjsx file. If you're up for it, ES6 too! But I don't know what the timeline for this may be, so if the deadline is strict, cjsx is ok for now too.

File: app/classifier/index.cjsx [hunk 1]
```diff
@@ -268,10 +267,46 @@ Classifier = React.createClass
...
+        subjectYPos = parseFloat(@props.subject.metadata['#Y1'])
+        subjectTol = parseFloat(@props.subject.metadata['#Tol1'])
+        userXPos = @props.classification.annotations[0]?.value[0]?.x
+        userYPos = @props.classification.annotations[0]?.value[0]?.y
+        isWithinTolerance = @withinTolerance(userXPos, userYPos, subjectXPos, subjectYPos, subjectTol)
+
+        console.log(subjectClass)
+        console.log(userMadeAnnotation)
+        console.log(@props.classification.annotations?.length)
+        console.log(@props.classification.annotations[0])
```
Comment (srallen) [inline app/classifier/index.cjsx]: The `console.log`s should be removed.

**Comment by srallen** [comment]
Are these messages specific to a single project? If so, we should consider what we're doing on [L310](https://github.com/hughdickinson/Panoptes-Front-End/blob/f5ad7c591306a988832a528a07a565a73594c198/app/classifier/index.cjsx#L310) to make sure this doesn't break another project's summary that doesn't have that kind of subject metadata. Though, I'm not convinced this kind of messaging should live in the subject metadata to start.

**Comment by camallen** [comment]
see #2834 for more details on what this is trying to achieve. 

Also the subject metadata is the ideal place to hold the ref for the summary information. It's subject specific information that can be re-used across projects. However hard coding language specific strings in there won't work. We would probably have to add workflow content references in there or something of the like for translations on this content to be supported (not that we currently are doing translations properly).

File: app/classifier/index.cjsx [hunk 2]
```diff
@@ -17,7 +17,7 @@ workflowAllowsFlipbook = require '../lib/workflow-allows-flipbook'
 workflowAllowsSeparateFrames = require '../lib/workflow-allows-separate-frames'
 WorldWideTelescope = require './world_wide_telescope'
 MiniCourseButton = require './mini-course-button'
-GridTool = require './drawing-tools/grid'
```
Comment (eatyourgreens) [inline app/classifier/index.cjsx]: Is this line supposed to be deleted?

File: app/classifier/index.cjsx [hunk 3]
```diff
@@ -406,11 +425,6 @@ Classifier = React.createClass
     @props.classification.update 'annotations'
 
   completeClassification: ->
-    currentAnnotation = @props.classification.annotations[@props.classification.annotations.length - 1]
-    currentTask = @props.workflow.tasks[currentAnnotation?.task]
-    currentTask?.tools?.map (tool) =>
-      if tool.type is 'grid'
-        GridTool.mapCells @props.classification.annotations
```
Comment (eatyourgreens) [inline app/classifier/index.cjsx]: Are these lines supposed to be deleted too?

File: app/classifier/index.cjsx [hunk 4]
```diff
@@ -268,10 +268,29 @@ Classifier = React.createClass
         </p>}
     </div>
 
+  withinTolerance: (userX, userY, metaX, metaY, tolerance) ->
+    distance = Math.sqrt((userY - metaY)**2 + (userX - metaX)**2)
+    isWithinTolerance = distance < tolerance
+    return isWithinTolerance
+
```
Comment (eatyourgreens) [inline app/classifier/index.cjsx]: Looks like this has moved into the feedback component, so it isn't needed here any more?

**Comment by hughdickinson** [comment]
@eatyourgreens You are correct about the last. Not sure how the others got changed, but I didn't do it deliberately.

File: app/classifier/tasks/metadata-based-feedback.cjsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
+React = require 'react'
+
+module.exports = React.createClass
+  displayName: 'MetadataBasedFeedback'
+
+  withinTolerance: (userX, userY, metaX, metaY, tolerance) ->
+    distance = Math.sqrt((userY - metaY)**2 + (userX - metaX)**2)
+    isWithinTolerance = distance < tolerance
+    return isWithinTolerance
```
Comment (srallen) [inline app/classifier/tasks/metadata-based-feedback.cjsx]: This is just an fyi about coffeescript: returns are implicit, so this line can just be `isWithinTolerance`

File: app/classifier/tasks/metadata-based-feedback.cjsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+  withinAnyTolerance: (userX, userY) ->
+    for key, value of @props.subject?.metadata
+      if key.indexOf(@props.metaSimCoordXPattern) == 0
```
Comment (srallen) [inline app/classifier/tasks/metadata-based-feedback.cjsx]: Generally in javascript, `===` should be used. In coffeescript, `is` can also be used. [Here's a good summary about why](http://stackoverflow.com/questions/359494/which-equals-operator-vs-should-be-used-in-javascript-comparisons)

**Comment by srallen** [comment]
We need confirmation that the same success and failure messages for sims/duds will be used for Space Warps in #2834 to avoid redundancy in the subject metadata like we discussed.

**Comment by hughdickinson** [comment]
Thanks for all the code review. I plan to implement the suggestions when I have time. This may take a couple of months though.

**Comment by hughdickinson** [comment]
Specific coding suggestions have been implemented and tested. Ready for another review depending upon whether the whole approach has been deemed inappropriate in the interim.

**Comment by srallen** [comment]
@hughdickinson Could you rebase to resolve the merge conflicts? @eatyourgreens do you have time to review this?

File: app/classifier/index.cjsx (hunk 3 above)
Comment (eatyourgreens) [inline app/classifier/index.cjsx]: I'm still wondering why these lines have been deleted. Would a rebase fix this?

**Comment by srallen** [comment]
@hughdickinson Could you set this PR to allow changes to be made by the maintainers? I believe you should see this option in the right sidebar of this PR. This will allow us to finish up the conflict resolution and rebase before merging.

**Comment by srallen** [comment]
Superceded by #3202.

**Pull request closed without merging**